compiler/damlc/daml-stdlib-src/DA/Internal/Desugar.daml

```diff
@@ -39,12 +40,30 @@ data PreConsuming t = PreConsuming {}
 data Consuming t = Consuming {}
 data PostConsuming t = PostConsuming {}
 
-class Implements t i where
+class HasInterfaceTypeRep i where
+  interfaceTypeRep : i -> TypeRep
+
+class HasToInterface i t where
   toInterface : t -> i
+
+class HasFromInterface t i where
   fromInterface : i -> Optional t
-  toInterfaceContractId : ContractId t -> ContractId i
-  fromInterfaceContractId : ContractId i -> Update (Optional (ContractId t))
-  interfaceTypeRep : t -> TypeRep
+
+type Implements t i =
+  ( HasInterfaceTypeRep i
+  , HasToInterface i t
+  , HasFromInterface t i
```

Comment:
I know I suggested it but, I really don't like the swapped order of arguments. I think it's going to cause confusion.

If we want `toInterface @Token` to work, we should probably just go with defining `_toInterface` in the typeclass and defining `toInterface` as a standalone function that invokes `_toInterface`.

Reply:
you're right, I was a bit confused myself on the `ghc` side, since the functions that generate the instances take two `LHsType`s and one needs to be mindful of the order. I think there's still value in splitting the `Implements` class, to be consistent with `Template`, but I'll keep the `t i` order for the single-method classes.

Reply:
> Ah, wait, this isn't how you make a constraint tuple, it's:

yes, sorry for the confusion, I fixed-up this commit. That's what I had before I switched to the no-method class